pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix

```
@@ -0,0 +1,177 @@
+#
+# preBuildAndTest and some small other bits
+# taken from https://github.com/tcdi/pgx/blob/v0.4.5/nix/extension.nix
+# which uses MIT License with the following license file
+#
+# MIT License
+#
+# Portions Copyright 2019-2021 ZomboDB, LLC.
+# Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>.
+# All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+{ lib
+, cargo-pgx
+, pkg-config
+, runCommand
+, rustPlatform
+, rustc
+, rustfmt
+, stdenv
+, Security
+}:
+
+#
+# The idea behind: Use it mostly like rustPlatform.buildRustPackage and so
+# we hand most of the arguments down.
+#
+# The only additional argument is `postgresql`.
+# You need to assign the targeted postgresql package to this argument.
+#
+
+{ buildAndTestSubdir ? null
+, buildType ? "release"
+, buildFeatures ? [ ]
+, cargoBuildFlags ? [ ]
+, postgresql
+, ...
+} @ args:
+
+assert lib.asserts.assertMsg ((args.installPhase or "") == "")
+  "buildPgxExtensions overwrites the installPhase, so providing one is useless";
+assert lib.asserts.assertMsg ((args.buildPhase or "") == "")
+  "buildPgxExtensions overwrites the buildPhase, so providing one is useless";
+
+let
+  maybeDebugFlag = if buildType != "release" then "--debug" else "";
+  maybeEnterBuildAndTestSubdir =
+    if buildAndTestSubdir != null then ''
+      export CARGO_TARGET_DIR="$(pwd)/target"
+      pushd ${buildAndTestSubdir}
+    '' else "";
+  maybeLeaveBuildAndTestSubdir = if buildAndTestSubdir != null then "popd" else "";
+
+  pgxPostgresMajor = builtins.head (lib.splitString "." postgresql.version);
+  preBuildAndTest = ''
+    export PGX_HOME=$(mktemp -d)
+    mkdir -p $PGX_HOME/${pgxPostgresMajor}
+
+    cp -r -L ${postgresql}/. $PGX_HOME/${pgxPostgresMajor}/
+    chmod -R ugo+w $PGX_HOME/${pgxPostgresMajor}
+    cp -r -L ${postgresql.lib}/lib/. $PGX_HOME/${pgxPostgresMajor}/lib/
+
+    ${cargo-pgx}/bin/cargo-pgx pgx init \
```

```suggestion
    cargo-pgx pgx init \
```

cargo-pgx must be in nativeBuildInputs or cross compiling does not work.